src/primitives/createFragment.ts

```diff
@@ -1,44 +1,94 @@
-import { getFragment } from "relay-runtime";
 import type { GraphQLTaggedNode } from "relay-runtime";
+import RelayRuntime from "relay-runtime/experimental";
 import type {
-	ArrayKeyType,
-	ArrayKeyTypeData,
 	KeyType,
 	KeyTypeData,
 } from "relay-runtime/lib/store/ResolverFragments";
-import { createResource } from "solid-js";
-import type { ResourceReturn } from "solid-js";
-import { unwrap } from "solid-js/store";
+import {
+	batch,
+	createComputed,
+	createMemo,
+	createResource,
+	onCleanup,
+} from "solid-js";
+import type { Accessor } from "solid-js";
 
+import { createStore, unwrap } from "solid-js/store";
 import { useRelayEnvironment } from "../RelayEnvironment";
-import { createFragmentNode } from "./createFragmentNode";
+import { type DataProxy, makeDataProxy } from "../utils/dataProxy";
+
+type FragmentResult<T> = {
+	data: T | undefined;
+	error: unknown;
+	pending: boolean;
+};
 
 export function createFragment<TKey extends KeyType>(
 	fragment: GraphQLTaggedNode,
-	key: TKey,
-): ResourceReturn<KeyTypeData<TKey>>;
+	key: Accessor<TKey>,
+): DataProxy<KeyTypeData<TKey>>;
 export function createFragment<TKey extends KeyType>(
 	fragment: GraphQLTaggedNode,
-	key: TKey | null,
-): ResourceReturn<KeyTypeData<TKey> | null>;
-export function createFragment<TKey extends ArrayKeyType>(
-	fragment: GraphQLTaggedNode,
-	key: TKey,
-): ResourceReturn<ArrayKeyTypeData<TKey>>;
-export function createFragment<TKey extends ArrayKeyType>(
+	key: Accessor<TKey | null>,
+): DataProxy<KeyTypeData<TKey> | null>;
+export function createFragment<TKey extends KeyType>(
 	fragment: GraphQLTaggedNode,
-	key: TKey | null,
-): ResourceReturn<ArrayKeyTypeData<TKey> | null> {
+	key: Accessor<TKey | null>,
+): DataProxy<KeyTypeData<TKey> | null> {
 	const environment = useRelayEnvironment();
-	const fragmentNode = getFragment(fragment);
-	return createResource(
-		createFragmentNode(
-			environment,
-			fragmentNode,
-			unwrap(key),
-			"createFragment()",
-		)[0],
-		// biome-ignore lint/suspicious/noExplicitAny: return type already enforced
-		(node) => node.data as any,
+
+	const source = createMemo(() => {
+		const k = unwrap(key());
+		return (
+			k && {
+				observable: RelayRuntime.observeFragment(environment, fragment, k),
+				promise: RelayRuntime.waitForFragmentData(environment, fragment, k),
```

Comment:
If you do both `observeFragment` and `waitForFragment` you are going to end up reading the data and subscribing twice which is likely pretty inefficient. I'm surprised you need the promise here, but I'm not that familiar with Solid.

Reply:
It was because I needed to get notified when the data for the fragment was fully ready, and `observeFragment` instantly passes data with `null`s through `next()` even if the fragment was waiting for a deferred payload. `waitForFragmentData` only gets resolved when the fragment data is fulfilled and that's exactly what I needed so I went this way.

Reply:
You should be able to use reuse the `observeFragment` observable and wait until the first `ok` event.

Reply:
Take a look at the implementation of `waitForFragmentData` for an idea of how to do this.

Reply:
As I mentioned before, it instantly passed `ok` with nulls 😅 Maybe that's unexpected...

Reply:
Ah, I had missed that. Do you have a repro?

Reply:
> Take a look at the implementation of waitForFragmentData for an idea of how to do this.

Wait, actually the implementation is almost identical to what I tried 🤔 let me investigate more...

Reply:
It actually works well 😅 maybe I forgot to check `status === "ok"` before...